### ISSUE TITLE
cylc graph (readability): node labels always white

### ISSUE THREAD

**Issue by sadielbartholomew** [cylc/cylc#2965]
A user notified us it was not easy to comprehend graphs within ``cylc graph`` due to the text, shapes & arrows being white & their light (default) background.

Upon investigation, these elements appear white despite the system e.g. GNOME theme used & context, e.g. even text inside light coloured task boxes. By contrast, there is no such problem with ``gcylc``, where text is displayed in a light, dark or middling shade based on the colour of the immediate background.

I suspect in the system GTK theming revamp, #2842, the ``cylc graph`` utility text was, perhaps indirectly, hard-coded to appear white. Apologies, as looking back, ``gcylc`` appears to have been thoroughly reviewed, whereas ``cylc graph`` was overlooked (at least by myself, as one reviewer :grimacing:).

#### Example screenshots

From my RHEL7 desktop with ``GNOME 2.28.2``:

| ![black_theme_1](https://user-images.githubusercontent.com/30274190/53192457-6a24a180-3606-11e9-835a-146ee34fa256.png) | ![white_theme_1](https://user-images.githubusercontent.com/30274190/53192458-6a24a180-3606-11e9-94ae-376deb5b0bf1.png) |
|:---:|:---:|
| **White on dark** (my own custom theme): fairly legible except for task node text | **White on light** (default theme *System*): unreadable in most components |

**Comment by sadielbartholomew** [comment]
... Actually in playing around to find a workaround for now, I recalled that displayed themes can become muddled when there is a change of the system (GNOME) theme while a GUI window is open, where the theme will then become correct after re-opening the window.

So this may not be a (Cylc) problem & if that emerges to be the case I will close this; while I do not see the behaviour now with window closing & opening on my RHEL6 environment with my setup, I have emailed the user to check the context in which they observed it.

**Comment by hjoliver** [comment]
I'll give this a test too when I have a chance.  I think I would have tested `cylc graph` with both light and dark themes, as well as `cylc gui`, but I'm not 100% sure.

**Comment by sadielbartholomew** [comment]
> displayed themes can become muddled when there is a change of the system (GNOME) theme while a GUI window is open, where the theme will then become correct after re-opening the window

As a brief update, I looked into this on RHEL7, where it was originally seen, & could not recreate it on there either. I checked in with the user that it was not simply the issue above, & they confirmed it was not. I have enquired as to whether it may just be a specific suite with something forcing node & edge attributes to white by hardcoding under ``[visualization]`` in the ``suite.rc``. I'll report any new info as I receive it here.

**Comment by sadielbartholomew** [comment]
I now suspect this is, in fact, a bug, because:

* the original observer has pointed me to some suites they were seeing it with & there was not anything under ``[visualization]`` that would hard-code all nodes and edges to one consistent colour, nor any reference to white or an equivalent colour code for;
* another Met Office user has observed the issue (independently), though only for the node labels, not edges or node shape outlines. I have changed the title here accordingly to reflect this probable scope.

It has also been noted, & confirmed by myself, that there is a distinct issue which was observed in an attempt to find a workaround. I will register it under this Issue simply because it also relates to theming & style & it seems overkill to open a new Issue. Namely, the ``[visualization] -> default node attributes`` item is not applying any set ``fillcolour=<colour>``, though it still recognises other ``key=value`` pairs such as ``style=filled`` & ``fillcolour=<colour>`` is still applied on a per-family basis under ``[[node attributes]]``.

**Comment by hjoliver** [comment]
Yeah, I've also noticed a problem with defaults coming in via `default node attributes`.  We should probably deprecate that and tell users just to style the root family, which seems to work fine (I suppose we should equally be able to make the `default` setting have exactly the same effect as styling root, but there may be some difficulty with that that I don't recall at the moment)... I haven't had time to take a look yet ...

**Comment by matthewrmshin** [comment]
Can we try putting in an automatic upgrade? E.g.: if `[visualization]default node attributes` is previously set but `[visualization][[root]]node attributes` is not, then we set `[visualization][[root]]node attributes` to the value of `[visualization]default node attributes`.

**Comment by hjoliver** [comment]
Yeah, let's try that.

**Comment by matthewrmshin** [comment]
I am trying to repeat this.
* Launch `cylc graph ...`
* Launch RHEL7's `gnome-tweaks`.
* Change from light to dark theme (or vice versa).
* Get dark-on-dark (or light-on-light) on `cylc graph ...`

However, the *Refresh* button can be used to force a re-draw - which fixes the issue. Or is there a signal/event the program can detect on theme change so it can do a re-draw?

**Comment by sadielbartholomew** [comment]
Thanks for looking into this, @matthewrmshin.

> Change from light to dark theme (or vice versa).

I was not sure originally, & I am still not convinced, that this is just a problem arising from theme changes, as described in https://github.com/cylc/cylc/issues/2965#issuecomment-466375154. In particular, it seems to emerge for particular suites, though I have yet had time to explore any characteristics shared by those.

I will continue to investigate it fairly soon, however.

> is there a signal/event the program can detect on theme change so it can do a re-draw?

If it is just an issue with a theme change, it is strictly not our problem, so we can close this Issue & mark it as a '*non-cylc bug*', but this idea seems a good one if possible. It is not clear to users that this is not an issue on our end so we really want to bypass it in a way like this so as not to give any bad impressions. I'll look into this too.

**Comment by hjoliver** [comment]
I've just tested this, for a suite with no visualization settings, and it works fine for me (Gnome 3) so long as you refresh the view after changing desktop theme:

Light theme:
![light](https://user-images.githubusercontent.com/2362137/55110096-2ad4fd80-513c-11e9-8284-1e499d4dcdec.png)

Dark theme, before refresh:
![light-dark](https://user-images.githubusercontent.com/2362137/55110142-45a77200-513c-11e9-8e37-f2b8e1346fd0.png)

Dark theme, after refresh:
![dark](https://user-images.githubusercontent.com/2362137/55110160-4dffad00-513c-11e9-80ca-7c29556d48fd.png)

**Comment by hjoliver** [comment]
If this only happens when a user changes theme (which is presumably a rare event) I don't think this is a major issue.  Although if there is some "theme change" event we can catch and trigger a redraw then we could do that.

@sadielbartholomew - does your example above have any `[visualization]` settings?  I wonder if the `default node attributes` settings are causing the problem somehow (as per speculation above).

**Comment by hjoliver** [comment]
Hmmm.  If I set a default fill color:
```
   [visualization]                                                             
      default node attributes = "style=filled", "fillcolor=yellow"  
```
Then I think we set the font color to be the same as the theme background color, otherwise (no fill) it is the theme foreground color.  This is a reasonable thing to do so long as the fill color isn't too light. To do this properly we need an automatic way to determine a font color that contrasts strongly with a given background color.

Dark theme:
![dark-yellow](https://user-images.githubusercontent.com/2362137/55110793-b4d19600-513d-11e9-84d5-5afb8082b440.png)
 
Light theme:
![light-yellow](https://user-images.githubusercontent.com/2362137/55110801-bb600d80-513d-11e9-8a40-4e002150ffe1.png)

(Then as mentioned above, `default node attributes` doesn't work if I try to set a specific fontcolor, but styling root does.)

**Comment by hjoliver** [comment]
Here's what I think is the right thing to do:
- use theme background as default fill color
- use theme foreground as default edge color, node border color, and font color
- if a user sets a non-default fill color, it is their responsibility to also set a font color that contrasts with their custom fill color.  The same would be required if you set a non-default page background color in any application.

I'll prepare a PR to stop the manipulation of font color according to whether or not a default fill color is set (which results in the white-on-yellow problem above). Then we should also fix the the `default node attributes` issues...

**Comment by sadielbartholomew** [comment]
> @sadielbartholomew - does your example above have any [visualization] settings? I wonder if the default node attributes settings are causing the problem somehow (as per speculation above).

@hjoliver, my initial Issue comment example was actually just the case arising from a switch of theme without refreshing the GUI window, so is not in fact an instance of the Cylc ``suite.rc`` problem. Sorry, I should have removed it accordingly when I realised that, & I will do that now.

It was (a few) users who noticed the problem of the unreadable node labels etc.*, as outlined in the bullet points in https://github.com/cylc/cylc/issues/2965#issuecomment-467559637, though I was unable to devote much time to extracting the relevant info from them so as to pinpoint or rule out possible root causes. I can still do that, but it might be best to wait to see if your PR addressing this Issue seems to solve the problem for them?

(*It did seem that their graphs were unreadable only for certain suites, which does imply ``[visualisation]`` settings were the origin, but I did check they hadn't set any such settings that should have led to the node fill and font being the same or similar, so there was something *wrong* with the processing of such settings, as opposed to *correct* hard-coding leading to it. )

**Comment by sadielbartholomew** [comment]
This should be resolved by #3047, though

> it was not easy to comprehend graphs within cylc graph due to the text, shapes & arrows being white & their light (default) background

was observed by some Met Office users under unknown circumstances, so I can't tell for sure. Certainly the:

> problem with defaults coming in via default node attributes

has been fixed. We could always re-open or create a new Issue if users notice further problems.
